sympy/core/numbers.py

```diff
@@ -1691,6 +1691,49 @@ def limit_denominator(self, max_denominator=1000000):
         f = fractions.Fraction(self.p, self.q)
         return Rational(f.limit_denominator(fractions.Fraction(int(max_denominator))))
 
+    def limit_num_denom(self, a, b = None):
+        """Closest Rational to self with numerator at most a
+           and denominator at most b if not None else Closest
+           Rational to self with denominator at most a.
+
+        Examples
+        ========
+
+        >>> from sympy import Rational
+        >>> Rational(1819831678027117, 9007199254740992).limit_num_denom(10**5)
+        11003/54459
+        >>> Rational(1819831678027117, 9007199254740992).limit_num_denom(10**3, 10**5)
+        376/1861
+
+        """
+        if a < 0:
+            raise AssertionError("Value should be greater than 0")
+
+        if self == 0:
+            return self
+        elif self < 0:
+            return -(-self).limit_num_denom(a, b)
+
+        if b is not None:
+            if b < 0:
+                raise AssertionError("Value should be greater than 0")
+            if self*b < 1:
+                return self.limit_denominator(b)
+            if b == 1:
+                return Integer(min(a, max(1, int(self))))
+            if a == 1:
+                if self >=1:
+                    return self.limit_num_denom(1, 1)
+                return self.limit_num_denom(1, int(1/self))
+            if self >=1:
+                b = min(a, b)
+            else:
+                a = min(a, b)
+            d_use = min(int(a/self), b)
```

Comment:
It may be more efficient to write this as `min((a*self.q)//self.p), b)` to just use the standard integer operations. (I think that `int` does the same as `//`)

Reply:
Nah, `int` and `//` behaves differently for negative numbers, but it would still be better to use `self.q` and `self.p` rather than relying on division by a `Rational`. So `min(int(a*self.q/self.p), b)` (and similar above in the other comment).

Reply:
The case with negative numbers has been handled previously so i think can we go ahead with `//` ?

Reply:
Yes, forgot to mention that it may still be possible. (This is where tests for all cases are handy, easy to see if one breaks anything by changing it.)